mmaction/core/multi_grid_utils.py

```diff
@@ -31,7 +29,7 @@ def modify_subbn3d_num_splits(logger, module, num_splits):
             new_split_bn = nn.BatchNorm3d(
                 child.num_features * num_splits, affine=False).cuda()
             new_state_dict = new_split_bn.state_dict()
-            for param_name, param in child.split_bn.state_dict().items():
+            for param_name, param in child.bn.state_dict().items():
```

Comment:
The implementation of PySlowfast also inheritate the num_batch_tracked of last split_bn. Is there any performance drop because of this.

Reply:
To be honest I haven't got time to test it yet. I have noticed that in Pyslowfast, they will save it but not load it as this if statement ```
            if (
                len(model_blob_shape) == 1
                and len(c2_blob_shape) == 1
                and model_blob_shape[0] > c2_blob_shape[0]
                and model_blob_shape[0] % c2_blob_shape[0] == 0
            ):
```
will filter it out.

Besides, when reading the source code of Pytorch, I found that `num_batch_tracked` will be used only when you specify the momentum to be None, check [this](https://github.com/pytorch/pytorch/blob/10c4b98ade8349d841518d22f19a653a939e260c/torch/nn/modules/batchnorm.py#L64) out. So I believe this buffer will not cause any error.

Reply:
You are welcomed to test this branch, before the latest commit 334955b, the test accuracy will be around 5 point lower than vanilla slowfast in our repo.

Reply:
Following your setting, the train top-1 accuracy could achieve 72%, but the test accuracy is very low. Have you encounter this problem?

Reply:
Did you use the latest commit? I got a top1_acc: 0.6771, top5_acc: 0.8763 at test set. Still 5 points lower, but I haven't tested 334955b yet.

Reply:
I still cannot reproduce the performance using 4*8 tesla V100 cards. The base lr is 4x than your setting.

Reply:
Would you please share your config and the test accuracy?

Reply:
I only change the learning rate to 0.4, and the test accuracy is about 0.72 top-1 acc.

Reply:
Then it works! The vanilla slowfast top-1 is also ~72.

Reply:
Then it is aboud 3 point lower than vanilla slowfast 8x8? Have you ever trained an 8x8 slowfast with stepwise learning rate? May stepwise lr will cause some loss in accuracy(cosine learning rate yield around 75 top-1)

Reply:
Besides, when testing, it is better use 'average-clips=prob', accuracy may get a gain of 0.2.

Reply:
May stepwise lr will cause some loss in accuracy. Have you tried it?

Reply:
Nah...Normally speaking consine lr is better for such large model(In PySlowFast the accuracy shows a minor drop of 0.2). However, lr and 'averagep-clips' are minor problems, a gap like 3~4 point still means something, like bugs.

Reply:
What's more, I do not have warmup now. This will also cause some minor drop. I think a better way is to train a stepwise baseline first and compare with it, rather than compare with the other slowfast trained with a lots of whisle and bells.

Reply:
Could you please test the checkpoint indicated in the `best.json`? I noticed last time when I train the model, the best is not the latest, but at about 330 epoch.

Reply:
I have test all checkpoints, but it's still much lower.

Reply:
Oh, my bad, here is a thing: you need to modify mmcv:
```
@@ -90,6 +90,7 @@ class EpochBasedRunner(BaseRunner):
         assert len(data_loaders) == len(workflow)

         self._max_epochs = max_epochs
+        self.call_hook('before_run')
```
I bet your latest checkpoint is epoch_239 rather than 358.

Reply:
I have noticed this bug and fixed it before training.

Reply:
Umm, what's your best accuracy? Maybe you can train a stepwise slowfast for reference.

Reply:
Have you tried training with PySlowFast before? I noticed that the best epoch is always earlier than finetune stage(around 325 or 330). This confused me a lot, seems like overfitting to the distribution of short-cycle(as this long-cycle is identical to vanilla one). Maybe turn off short-cycle at the last epochs is better, but I haven't seem any code in PySlowfast to do so.

Reply:
top1_acc	0.7059
top5_acc	0.8949
That's my highest accuracy, top1 ~5 points lower, top5 ~3 points lower.

Reply:
I have tried BN3d which could achieve higher performance (1%-2%) than subBN3d.

Reply:
That's strange...I think the major bug is still in SubBN3d, but have no idea where it is. We cannot merge this PR until it achieve the same accuracy as vanilla slowfast.

Reply:
Hi @donnyyou,

Would you kindly give us your best performanced checkpoint (that one you use BN3d), or raise a PR by yourself? We may test and release it if it is close enough.

Reply:
I trained the 8x8x1 slowfast model on Kinetics400 with cosine LR for 239 epochs, but get 1.5 point lower than the result you report!

Reply:
Have you tried to train with 239 epochs rather than 256 epochs?

Reply:
No, I haven't tested it yet. The augmentation is slightly different and may need a longer epoch.

Reply:
Hi! Could you share the kinetics400 dataset with me? The dataset I used seemingly has something wrong.

Reply:
Have you tried our preparation scripts? It is too big to share...

Reply:
With your preparation scripts, there are only 220034 lines for train.csv.
But https://raw.githubusercontent.com/activitynet/ActivityNet/master/Crawler/Kinetics/data/kinetics-400_train.csv has 246535 lines. I wonder which is the correct one?

Reply:
According to your training log, your train dataset has near 24W samples.

Reply:
Could you share your annotations for downloading Kinetics400?

Reply:
Yeah, 240436 and validation 19796 is the standard k400 for all experiments used in our repo. Try this [train anno](https://drive.google.com/file/d/1WraT4uwH6uEJ75X_jDxlfo20UNXRiIO2/view?usp=sharing) [test anno](https://drive.google.com/file/d/17P1vjTIXqD-8_L0oM5aw6Y54O8zVLY0G/view?usp=sharing).

Reply:
Thanks for your help! I have achieved top-1 75.2 with short-side 256 trainset using multi-grid sampling strategy.

Reply:
That's great! Could you please share us your checkpoint and logs?

Reply:
I will share you these in this week!

Reply:
Hi @donnyyou, do you have the ready checkpoint and logs now? You may share via GoogleDrive or other cloud storages.

Reply:
Sorry! I'm busy these days. The checkpoints and logs can't be shared because of the rules of our company. You could simply implement the better results (2-3%) with precise BN according to the setting of PySlowFast.

Reply:
Also I have changed the augmentations with random short side in [256, 320], then random crop the target area as multigrid scheduling.

Reply:
At last, the interval for frames sampling is random in [2, T/t * 2], not a constant of T/t * 2. This will result in better convergence.

Reply:
@SuX97 @dreamerlin

Reply:
> Also I have changed the augmentations with random short side in [256, 320], then random crop the target area as multigrid scheduling.

This will bring a little performance gain.

Reply:
Thanks for reporting, @donnyyou! To clarify, the diff between this branch and yours is not in the model or multi-grid schedule itself, but lies in the data pipeline that is more random like PySlowFast?

Reply:
Precise BN is the most important. @SuX97

Reply:
How much gain does precise BN bring according to your experiment? But it is unfair to compare with the baseline if apply precise BN at test time(other checkpoints in our repo are tested with vanilla bn).

Reply:
Precise BN is the essential part of multi-grid training, which brings about 2% performance gain. I guess the main reason is that the various sizes of  input during multi-grid sampling leads to inaccurate BN stats.

Reply:
BTW, do you mean by performing precise bn when testing, or performing precise bn when training (override the stats in all bns after each epoch)?

Reply:
during training.